parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+//! a deposit, as it is expected to be managed by a trusted collective.
+//!
+//! Content types:
+//!
+//! - Collective [charter](pallet::Charter): A single document (`Cid`) managed by
+//!   [CharterOrigin](pallet::Config::CharterOrigin).
+//! - Collective [announcements](pallet::Announcements): A list of announcements managed by
+//!   [AnnouncementOrigin](pallet::Config::AnnouncementOrigin).
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+use frame_support::{traits::schedule::DispatchTime, BoundedVec};
+use sp_core::ConstU32;
+use sp_std::prelude::*;
+
+/// IPFS compatible CID.
+// worst case 2 bytes base and codec, 2 bytes hash type and size, 64 bytes hash digest.
+pub type Cid = BoundedVec<u8, ConstU32<68>>;
+
+/// The block number type of [frame_system::Config].
+pub type BlockNumberFor<T> = <T as frame_system::Config>::BlockNumber;
+
+/// [DispatchTime] of [frame_system::Config].
+pub type DispatchTimeFor<T> = DispatchTime<BlockNumberFor<T>>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	/// The module configuration trait.
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The origin to control the collective announcements.
+		type AnnouncementOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin to control the collective charter.
+		type CharterOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The maximum number of announcements.
+		///
+		/// NOTE: Benchmarks need to be re-run if this changes.
+		#[pallet::constant]
+		type MaxAnnouncementsCount: Get<u32>;
+
+		/// Weight information needed for the pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// The announcement is not found.
+		MissingAnnouncement,
+		/// Number of announcements exceeds `MaxAnnouncementsCount`.
+		TooManyAnnouncements,
+		/// Cannot expire in the past.
+		InvalidExpiration,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		/// A new charter has been set.
+		NewCharterSet { cid: Cid },
+		/// A new announcement has been made.
+		AnnouncementAnnounced { cid: Cid, maybe_expire_at: Option<T::BlockNumber> },
+		/// An on-chain announcement has been removed.
+		AnnouncementRemoved { cid: Cid },
```

Comment:
Curiosity killed the cat, but everyone is going to want to know why an announcement has been removed. Would be cool to have an enum of Outdated, Retracted, Other as to the announcement removal reason.